### PR TITLE
improvement: resolve project app names set using a module attribute

### PR DESCRIPTION
**Pull request by zachallaun**
This PR also fixes two bugs in `Common.expand_literal/1`. Previously, the following would be returned:

```elixir
iex(1)> ":foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Igniter.Code.Common.expand_literal()
{:ok, {:__block__, [trailing_comments: [], leading_comments: [], line: 1, column: 1], [:foo]}}
iex(2)> "@foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Igniter.Code.Common.expand_literal()
{:ok,
 {:@, [trailing_comments: [], leading_comments: [], line: 1, column: 1],
  [{:foo, [trailing_comments: [], leading_comments: [], line: 1, column: 2], nil}]}}
```

With this PR:

```elixir
iex(1)> ":foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Igniter.Code.Common.expand_literal()
{:ok, :foo}
iex(2)> "@foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Igniter.Code.Common.expand_literal()
:error
```

### Contributor checklist

- [x] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

### CODE REVIEW

File: lib/igniter/code/module.ex
Comment (zachallaun) [inline lib/igniter/code/module.ex:51]: This seemed like a potentially useful API (given a node, find the containing defmodule), but I didn't want to shoehorn it into this PR. If you'd prefer this not be here, I'm happy to make it a private function where it's used.

File: lib/igniter/project/application.ex
Comment (zachallaun) [inline lib/igniter/project/application.ex:86]: Oops, cruft from debugging, will fix and rebase.

File: lib/igniter/code/module.ex
Comment (zachdaniel) [reply lib/igniter/code/module.ex:51]: Seems very useful indeed! Go for it 😁

Comment (zachallaun) [reply lib/igniter/code/module.ex:51]: Just to be clear: Should I make it public and document it in this PR, or in a separate PR after this is merged?

File: lib/igniter/project/application.ex
Comment (zachdaniel) [inline lib/igniter/project/application.ex:54]: I don't think that we can rely on this unfortunately. Igniter is predicated on the fact that other things may have written to this zipper before this function was called, and so may not have line number info.  i think we have to "move out to the defmodules scope" and then move up until we find the module attribute declaration.

File: lib/igniter/code/module.ex
Comment (zachdaniel) [reply lib/igniter/code/module.ex:51]: 😅sorry, in this PR is good 😀

File: lib/igniter/project/application.ex
Comment (zachallaun) [reply lib/igniter/project/application.ex:54]: Ah, yes, you're right. If that's the case, I won't need the `Module.move_up_to_defmodule/1` helper. If it's alright with you, I'd prefer to remove it from this PR and I can submit a followup shortly after.

File: lib/igniter/code/module.ex
Comment (zachallaun) [reply lib/igniter/code/module.ex]: See [here](https://github.com/ash-project/igniter/pull/111#discussion_r1817875500); plan to do this in another PR since this function isn't necessary for this PR anymore.

File: lib/igniter/project/application.ex
Comment (zachdaniel) [inline lib/igniter/project/application.ex:56]: Could this be replaced by a `move_left` function with a predicate?

**Comment by zachdaniel** [comment]
Looking great. One small question added. Also curious if you'd rather get the other one in before this.

**Comment by zachdaniel** [comment]
We can ignore the CI issue, that is a misconfiguration in our CI that I need to address.

File: lib/igniter/project/application.ex
Comment (zachallaun) [reply lib/igniter/project/application.ex:56]: It could, but I don't think `move_left` exists yet, unless I somehow missed it.

We could add `move_left` in the other PR, get it merged in, then update this PR to use it.

Comment (zachdaniel) [reply lib/igniter/project/application.ex:56]: I'd vote for that I think :) Making more ergonomic (and idiomatic) public APIs to let people build with igniter is high priority at the moment, so I'm always looking for opportunities for that.

**Comment by zachallaun** [comment]
Marked as a draft for now because this PR now relies on #113, which hasn't been merged yet, so tests are now failing. Once it's merged, I'll rebase and mark as ready for review.

File: lib/igniter/project/application.ex
Comment (zachallaun) [reply lib/igniter/project/application.ex]: Definitely a nice improvement using `move_left`! 08db3e5ef1b48096135f88caf1a14ca25d98cb79

**Comment by zachallaun** [comment]
I believe this is ready to go pending a final review, @zachdaniel :)

File: lib/igniter/project/application.ex
Comment (zachallaun) [reply lib/igniter/project/application.ex:69]: Also, I'm realizing in looking at this that I should use the new `Common.move_upwards` that was implemented in the other PR.

Comment (zachdaniel) [reply lib/igniter/project/application.ex:69]: 🤔 thinking about it, do we necessarily need to check that we're in the module body? We can tell the difference between the AST that assigns to a module attribute and the AST for accessing it, right? So we could theoretically just `move_left/up` (a.k.a prev, right?) until we find assignment to that @attr?

Comment (zachallaun) [reply lib/igniter/project/application.ex:69]: 👍 Good idea, will refactor.

Comment (zachdaniel) [reply lib/igniter/project/application.ex:69]: I did make a conscious choice to treat everything is "down/right", i.e `move_to`. I think this simplifies a lot of the confusing parts of using zippers, and is 99% of the time what those writing generators/patchers will need to do. But it does mean that you can't just "reverse" most operations, so we will need conscious stuff like "move_back_to". I hate that name, so something else 😆

Comment (zachallaun) [reply lib/igniter/project/application.ex]: Refactored; good call on this, I do think it's more understandable.

**Review by zachdaniel — APPROVED** [review]

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**
